File: apps/api/internal/http/handlers/media.go
package handlers

import (
	"net/http"
	"strconv"

	"github.com/onnwee/artemis/apps/api/internal/service/nasa"
)

type MediaHandler struct {
	client *nasa.ImagesClient
	apod   *nasa.ApodClient
}

func NewMediaHandler(client *nasa.ImagesClient, apod *nasa.ApodClient) *MediaHandler {
	return &MediaHandler{client: client, apod: apod}
}

func (h *MediaHandler) Search(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("q")
	if query == "" {
		query = "artemis"
	}
	mediaType := r.URL.Query().Get("type")
	yearStart, _ := strconv.Atoi(r.URL.Query().Get("year"))
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))

	result, err := h.client.Search(r.Context(), query, mediaType, yearStart, page)
	if err != nil {
		writeError(w, http.StatusBadGateway, "failed to search NASA media")
		return
	}

	writeJSON(w, http.StatusOK, result)
}

func (h *MediaHandler) Apod(w http.ResponseWriter, r *http.Request) {
	apod, err := h.apod.Fetch(r.Context())
	if err != nil {
		writeError(w, http.StatusBadGateway, "failed to fetch APOD")
		return
	}
	writeJSON(w, http.StatusOK, apod)
}

File: apps/web/src/api/media.ts
import { api } from "./client";
import type { MediaSearchResult } from "@/lib/types";

export interface Apod {
  title: string;
  explanation: string;
  url: string;
  hdurl?: string;
  media_type: string;
  date: string;
  copyright?: string;
}

export function searchMedia(
  query = "artemis",
  type?: string,
  year?: number,
  page?: number
) {
  const params = new URLSearchParams();
  params.set("q", query);
  if (type) params.set("type", type);
  if (year) params.set("year", String(year));
  if (page && page > 1) params.set("page", String(page));
  return api<MediaSearchResult>(`/api/v1/media?${params}`);
}

export function getApod() {
  return api<Apod>("/api/v1/apod");
}
